kong/runloop/balancer/ewma.lua

```diff
@@ -0,0 +1,255 @@
+end
+
+local function pick_and_score(self, address, k)
+  shuffle_address(address, k)
+  local lowest_score_index = 1
+  local lowest_score = get_or_update_ewma(self, address[lowest_score_index], 0, false) / address[lowest_score_index].weight
+  for i = 2, k do
+    local new_score = get_or_update_ewma(self, address[i], 0, false) / address[lowest_score_index].weight
+    if new_score < lowest_score then
+      lowest_score_index, lowest_score = i, new_score
+    end
+  end
+  return address[lowest_score_index], lowest_score
+end
+
+
+function ewma:getPeer(cacheOnly, handle, valueToHash)
+  if handle then
+    -- existing handle, so it's a retry
+    handle.retryCount = handle.retryCount + 1
+
+    -- keep track of failed addresses
+    handle.failedAddresses = handle.failedAddresses or setmetatable({}, {__mode = "k"})
+    handle.failedAddresses[handle.address] = true
+  else
+    handle = {
+        failedAddresses = setmetatable({}, {__mode = "k"}),
+        retryCount = 0
+    }
+  end
+
+  if not self.balancer.healthy then
+    return nil, balancers.errors.ERR_BALANCER_UNHEALTHY
+  end
+
+  -- select first address
+  local address
+  for addr, ewma in pairs(self.ewma) do
+    if ewma ~= nil then
+      address = addr
+      break
+    end
+  end
+
+  if address == nil then
+    -- No peers are available
+    return nil, balancers.errors.ERR_NO_PEERS_AVAILABLE, nil
+  end
+
+  local ip, port, host
+  while true do
+    -- retry end
+    if self.address_count > 1 then
+      local k = (tonumber(self.address_count) < PICK_SET_SIZE) and tonumber(self.address_count) or PICK_SET_SIZE
+      local filtered_address = {}
+  
+      for addr, ewma in pairs(self.ewma) do
+        if not handle.failedAddresses[addr] then
+          table_insert(filtered_address, addr)
+        end
+      end
+  
+      if #filtered_address == 0 then
+        ngx_log(ngx.WARN, "all endpoints have been retried")
+        ip, port, host = balancers.getAddressPeer(address, cacheOnly)
+        handle = {
+          failedAddresses = setmetatable({}, {__mode = "k"}),
+          retryCount = 0,
+          address = address,
+        }
+        return ip, port, host, handle
+      end
+
+      local score
+      if #filtered_address > 1 then
+        k = #filtered_address > k and #filtered_address or k
+        address, score = pick_and_score(self, filtered_address, k)
+      else
+        address, score = filtered_address[1] ,get_or_update_ewma(self, filtered_address[1], 0, false)
+      end
+      ngx_log(ngx_DEBUG, "get ewma score: ", score)
+    end
+    -- check the address returned, and get an IP
+
+    ip, port, host = balancers.getAddressPeer(address, cacheOnly)
+    if ip then
+      -- success, exit
+      handle.address = address
+      break
+    end
+
+    handle.failedAddresses[address] = true
+    if port ~= balancers.errors.ERR_DNS_UPDATED then
+      -- an unknown error
+      break
+    end
+  end
+
+  if ip then
+    return ip, port, host, handle
+  else
```

Comment:
blank line before `else`

Reply:
The `else` statement could be removed at all.